<a href="https://colab.research.google.com/github/Jiangjiang-111/-/blob/main/%E7%AC%AC%E4%BA%8C%E7%AB%A0%EF%BC%9A%E7%AC%AC%E4%BA%8C%E8%8A%82%E6%95%B0%E6%8D%AE%E9%87%8D%E6%9E%841_%E8%AF%BE%E7%A8%8B%EF%BC%88done!%EF%BC%89.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**复习：**在前面我们已经学习了Pandas基础，第二章我们开始进入数据分析的业务部分，在第二章第一节的内容中，我们学习了**数据的清洗**，这一部分十分重要，只有数据变得相对干净，我们之后对数据的分析才可以更有力。而这一节，我们要做的是数据重构，数据重构依旧属于数据理解（准备）的范围。

#### 开始之前，导入numpy、pandas包和数据

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
# 导入基本库
import numpy as np
import pandas as pd

In [4]:
# 载入data文件中的:train-left-up.csv
train_left_up = pd.read_csv("/content/drive/MyDrive/hands-on-data-analysis-master/第二章项目集合/data/train-left-up.csv")
train_left_up.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  
0                            Braund, Mr. Owen Harris  
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  
2                             Heikkinen, Miss. Laina  
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  
4                           Allen, Mr. William Henry

# 2 第二章：数据重构


### 2.4 数据的合并

#### 2.4.1 任务一：将data文件夹里面的所有数据都载入，观察数据的之间的关系

In [5]:
#写入代码
train_left_down = pd.read_csv("/content/drive/MyDrive/hands-on-data-analysis-master/第二章项目集合/data/train-left-down.csv")
train_right_down = pd.read_csv("/content/drive/MyDrive/hands-on-data-analysis-master/第二章项目集合/data/train-right-down.csv")
train_right_up = pd.read_csv("/content/drive/MyDrive/hands-on-data-analysis-master/第二章项目集合/data/train-right-up.csv")

In [6]:
#写入代码
train_left_down.head()

PassengerId  Survived  Pclass                                         Name
0          440         0       2       Kvillner, Mr. Johan Henrik Johannesson
1          441         1       2  Hart, Mrs. Benjamin (Esther Ada Bloomfield)
2          442         0       3                              Hampe, Mr. Leon
3          443         0       3                    Petterson, Mr. Johan Emil
4          444         1       2                    Reynaldo, Ms. Encarnacion

In [7]:
train_left_down.tail()

PassengerId  Survived  Pclass                                      Name
447          887         0       2                     Montvila, Rev. Juozas
448          888         1       1              Graham, Miss. Margaret Edith
449          889         0       3  Johnston, Miss. Catherine Helen "Carrie"
450          890         1       1                     Behr, Mr. Karl Howell
451          891         0       3                       Dooley, Mr. Patrick

In [8]:
train_right_up.head()

Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S
1  female  38.0      1      0          PC 17599  71.2833   C85        C
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S
3  female  35.0      1      0            113803  53.1000  C123        S
4    male  35.0      0      0            373450   8.0500   NaN        S

In [9]:
train_right_down.head()

Sex   Age  SibSp  Parch        Ticket    Fare Cabin Embarked
0    male  31.0      0      0    C.A. 18723  10.500   NaN        S
1  female  45.0      1      1  F.C.C. 13529  26.250   NaN        S
2    male  20.0      0      0        345769   9.500   NaN        S
3    male  25.0      1      0        347076   7.775   NaN        S
4  female  28.0      0      0        230434  13.000   NaN        S

【提示】结合之前我们加载的train.csv数据，大致预测一下上面的数据是什么

* train-left-up 是passengerID1~339(前半部分)的乘客，是否幸存、仓等级、姓名等数据(左侧列特征)
* train_left_down 是passengerID440~891（后半部分）的乘客.....（左侧列特征）
* train_right_up 是passengerID1~339的乘客，右侧列特征
* train_right_down 是passengerID440~891的乘客，右侧列特征

**concat** 连接数据集 

* 默认按行将多个DataFrame组合到一起的

* 必须指定ignore_index=True,即不保留所返回的原始行号

#### 2.4.2：任务二：使用concat方法：将数据train-left-up.csv和train-right-up.csv横向合并为一张表，并保存这张表为result_up

In [10]:
#写入代码
result_up = pd.concat([train_left_up,train_right_up],axis=1)
result_up.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

#### 2.4.3 任务三：使用concat方法：将train-left-down和train-right-down横向合并为一张表，并保存这张表为result_down。然后将上边的result_up和result_down纵向合并为result。

In [11]:
#写入代码
result_down = pd.concat([train_left_down,train_right_down],axis=1)
result = pd.concat([result_up,result_down],ignore_index=True)

In [12]:
result.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [13]:
result.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

#### 2.4.4 任务四：使用DataFrame自带的方法join方法和append：完成任务二和任务三的任务

**join** 连接

* data.join(other) 将一个DataFrame加入到当前DataFrame中
* data.join([ , ]) 将多个DataFrame加入到当前DataFrame中
* how: 指定连接方式，默认为left

| 连接方式 | 含义 | 表示 |
| ---- | ---- | ---- |
| inner |	内连 |	取行索引的交集
| outer	| 外连 |	取行索引的并集
| left | 左连 |	使用左边df的行索引
| right	| 右连 |	使用右边df的行索引

* on=" " 设置用于连接的列,默认为行连接（None） 注意：合并多个DataFrame时，只支持用DataFrame的行索引进行连接,on不能用

* 默认以左连接的方式进行合并，默认的连接列是DataFrame的行索引，并且，合并两个DataFrame时，两个DataFrame中不能有相同的列名

In [17]:
#写入代码
#将数据train-left-up.csv和train-right-up.csv横向合并为一张表，并保存这张表为result_up1
#将train-left-down和train-right-down横向合并为一张表，并保存这张表为result_down1。然后将上边的result_up和result_down纵向合并为result。

result_up1 = train_left_up.join(train_right_up)
result_down1 = train_left_down.join(train_right_down)
result1 = result_up1.append(result_down1)

In [18]:
result1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [19]:
result1.tail()

PassengerId  Survived  Pclass                                      Name  \
447          887         0       2                     Montvila, Rev. Juozas   
448          888         1       1              Graham, Miss. Margaret Edith   
449          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
450          890         1       1                     Behr, Mr. Karl Howell   
451          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
447    male  27.0      0      0      211536  13.00   NaN        S  
448  female  19.0      0      0      112053  30.00   B42        S  
449  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
450    male  26.0      0      0      111369  30.00  C148        C  
451    male  32.0      0      0      370376   7.75   NaN        Q

#### 2.4.5 任务五：使用Panads的merge方法和DataFrame的append方法：完成任务二和任务三的任务

In [22]:
#写入代码
#将train-left-down和train-right-down横向合并为一张表，并保存这张表为result_down1。然后将上边的result_up和result_down纵向合并为result。
result_up = pd.merge(train_left_up,train_right_up,left_index=True,right_index=True)
result_down = pd.merge(train_left_down,train_right_down,left_index=True,right_index=True)
result = result_up.append(result_down)
result.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

【思考】对比merge、join以及concat的方法的不同以及相同。思考一下在任务四和任务五的情况下，为什么都要求使用DataFrame的append方法，如何只要求使用merge或者join可不可以完成任务四和任务五呢？

#### 2.4.6 任务六：完成的数据保存为result.csv

In [23]:
#写入代码
result.to_csv("result.csv")

### 2.5 换一种角度看数据

#### 2.5.1 任务一：将我们的数据变为Series类型的数据

In [24]:
# 将完整的数据加载出来
text = pd.read_csv('result.csv')
text.head()
# 代码写在这里
unit_result=text.stack().head(20)
unit_result.head()


0  Unnamed: 0                           0
   PassengerId                          1
   Survived                             0
   Pclass                               3
   Name           Braund, Mr. Owen Harris
dtype: object

In [25]:
#将代码保存为unit_result,csv
unit_result.to_csv('unit_result.csv')

In [26]:
test = pd.read_csv('unit_result.csv')

In [27]:
test.head()

Unnamed: 0   Unnamed: 1                        0
0           0   Unnamed: 0                        0
1           0  PassengerId                        1
2           0     Survived                        0
3           0       Pclass                        3
4           0         Name  Braund, Mr. Owen Harris